pkg/clients/compute/agentpool/agentpool.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package agentpool
+
+import (
+	"github.com/crossplane/provider-azure/pkg/clients/compute/consts"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2020-03-01/containerservice"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/crossplane/provider-azure/apis/compute/v1alpha3"
+	azure "github.com/crossplane/provider-azure/pkg/clients"
+)
+
+// New returns an Azure AgentPool object from a AgentPool spec
+func New(c *v1alpha3.AgentPool) containerservice.AgentPool {
+	p := containerservice.AgentPool{
+		ManagedClusterAgentPoolProfileProperties: &containerservice.ManagedClusterAgentPoolProfileProperties{
+			VMSize:            containerservice.VMSizeTypes(c.Spec.NodeVMSize),
+			EnableAutoScaling: azure.ToBoolPtr(c.Spec.EnableAutoScaling),
+			// AgentPool APIs supported only for VMSS agentpools.
+			// For more information, please check https://aka.ms/multiagentpoollimitations
+			Type: containerservice.VirtualMachineScaleSets,
+			// Defaults override below if needed
+			Mode: containerservice.User,
+		},
+	}
+	if c.Spec.Mode != "" {
+		p.ManagedClusterAgentPoolProfileProperties.Mode = containerservice.AgentPoolMode(c.Spec.Mode)
+	}
+	if len(c.Spec.AvailabilityZones) > 0 {
+		p.AvailabilityZones = &c.Spec.AvailabilityZones
+	}
+	if c.Spec.NodeCount != nil {
+		p.ManagedClusterAgentPoolProfileProperties.Count = c.Spec.NodeCount
+	}
+	if c.Spec.MinNodeCount != nil {
+		p.ManagedClusterAgentPoolProfileProperties.MinCount = c.Spec.MinNodeCount
+	}
+	if c.Spec.MaxNodeCount != nil {
+		p.ManagedClusterAgentPoolProfileProperties.MaxCount = c.Spec.MaxNodeCount
+	}
+	if c.Spec.VnetSubnetID != "" {
+		p.VnetSubnetID = to.StringPtr(c.Spec.VnetSubnetID)
+	}
+	if len(c.Spec.NodeTaints) > 0 {
+		p.NodeTaints = &c.Spec.NodeTaints
+	}
+	return p
+}
+
+// NewProfile returns an Azure ManagedClusterAgentPoolProfile object from a AKSCluster spec
+func NewProfile(c *v1alpha3.AKSCluster) containerservice.ManagedClusterAgentPoolProfile {
+	nodeCount := int32(v1alpha3.DefaultNodeCount)
+	if c.Spec.NodeCount != nil {
+		nodeCount = int32(*c.Spec.NodeCount)
+	}
+	p := containerservice.ManagedClusterAgentPoolProfile{
+		Name:   to.StringPtr(consts.AgentPoolProfileName),
```

Comment:
AgentPoolProfile is subresource of AKSCluster and is required. Hardcoded AgentPoolProfileName was there before my patch. Do we really need changed this? And from my point of view, it would be better for users to hide AgentPoolName from AKSCluster spec.